### ISSUE TITLE
Add `--ext` flag and corresponding Node.js API option

### ISSUE THREAD

**Issue by ybiquitous** [stylelint/stylelint#6689]
### What is the problem you're trying to solve?

Splitted from #3411 
See also the discussion on PR #6617 for more details.

In brief, we have to specify glob patterns currently like this:

```shell
stylelint "**/*.css"
```

or

```shell
stylelint "src/**/*.{css,scss}"
```

This requires a lot of input letters and even knowledge of globs.

### What solution would you like to see?

Instead, this feature allows people to do the same thing like this:

```shell
stylelint .
```

(assuming `--ext` defaults to `.css`)

or

```shell
stylelint src --ext=.css,.scss
```

### Related

- #3411 
- #3860
- #4125

**Comment by silverwind** [comment]
Seconding the request. If stylelint is passed a directory, it will lint all files regardless of their extension. In our case, we had `.go` files in the same directory and we had to use `ignoreFiles: "**/*.go"` to exclude them.

Having a `--ext` option like eslint that defaults to `css` would be ideal.

**Comment by ybiquitous** [comment]
@silverwind Exactly; we will plan that `--ext` defaults to `.css` if omitted. You can see the related code and discussion on PR #6617.

**Comment by jeddy3** [comment]
I appreciate that a lot of work and discussion has gone into https://github.com/stylelint/stylelint/pull/6617. As @Mouvedia mentioned in https://github.com/stylelint/stylelint/issues/3411#issuecomment-1450529431, ESLint has moved away `--ext` for their [new configuration](https://eslint.org/docs/latest/use/configure/configuration-files-new). We typically try to align with ESLint's approaches and the direction they are taking as more people are familiar with that tool.

Should we pick up https://github.com/stylelint/stylelint/issues/3860 instead, and align the ignoring and resolution behaviour with ESLint's?

**Comment by Mouvedia** [comment]
It would incur a small performance cost to support #3411. It requires system calls to find all the configs that extend the current one, to merge the extensions.
At first, I included all 3—CLI, Node and config. I can add back `extensions` config property but Ill wait for a clear resolution on the matter here.

> Should we pick up https://github.com/stylelint/stylelint/issues/3860 instead, and align the ignoring and resolution behaviour with ESLint's?

Personally what Id advocate for is this:

- add `--ext`
- deprecate `ignoreFiles` and rename it to `ignores`
- add `files` config property

That what I was planning to do.

**Comment by ybiquitous** [comment]
It's nice timing to reconsider the related issues before shipping `--ext`.

Even if we have the `files: ["*.css"]` config, there may be a helpful case for `--ext`. For example:

```shell
# People might want to lint `.js` files regardless of the `files: ["*.css"]` config.
stylelint --ext .js src/
```

This use-case may be so rare, though. 😅 

However, for now, I'm not sure how complex the codebase will be by supporting both `files: []` config and `--ext`. I prefer a simpler way as possible, considering maintainability.

**Comment by silverwind** [comment]
Having all files defined in the config instead of the CLI would be welcome from my side. It always bothered me that part of the config is on the CLI and part in `.stylelintrc.*`.

But I agree that if eslint is to be followed, the `files` and `ignores` options should be mirrored and the config file to be reduced to just `stylelint.config.js`, dropping the `cosmicconfig` dependency in turn.

**Comment by ybiquitous** [comment]
Surely, it's a nice idea to make the relationship with CLI and the config more straightforward.

Regarding the dependency on `cosmiconfig`, I also guess that we could simplify the config system, but there should be many considerations and obstacles for the migration. We might also learn a lot from ESLint about this point.

---

About the rare case on my <https://github.com/stylelint/stylelint/issues/6689#issuecomment-1454269722>. I notice we could use just a glob pattern instead of `--ext`. See an example below:

```jsonc
// .stylelintrc.json
{
  "files": ["*.{css,scss}", "**/*.{css,scss}"]
}
```

```shell
# When we want to lint all files defined in the config (`.css` and `.scss`).
stylelint

# When we want to lint only `.scss` files.
stylelint "**/*.scss"

# Even when we want to lint `.js` files regardless of the config (`.css` and `.scss`).
stylelint "src/**/*.js"
```

In short, I guess we would no longer need the `--ext` flag. Instead, we would need to learn glob patterns, but it seems acceptable.

**Comment by Mouvedia** [comment]
Initially I picked #3411 because it was one of the oldest requests with a lot of 👍 .
The core of it was to add `extensions` config property. I was asked to drop it.
What was left was CLI and Node API. If we are to fulfil the original demand at least partially, `--ext` needs to remain.
Both #3411 and #6689 were _ready to implement_ or _wip_. Doing a 180 once the PR is ready to be merged will create discontentment.

Concerning ESLint RFC 9, you have to understand that their deprecation strategies works differently than ours.
Here's an example for rules:

> Since deprecated rules will never be removed, you can continue to use them indefinitely if they are working for you. However, keep in mind that deprecated rules will effectively be unmaintained.

We are planning to remove stylistic rules in the next major. ESLint currently supports `--ext` and even if they plan to remove it, it will still be supported for a while. If we are to align, to a certain extent, we should do it on the basis of the current version.

related issues: #4661 #4083

**Comment by ybiquitous** [comment]
@Mouvedia I appreciate your work. But I believe we should avoid potential problems if we notice them before releasing them.

We should consider both people's usability and our maintainability at a balance. If we quickly implement a feature and notice inconsistency later, we maintainers would hold headaches for a long time. It could lead to burnout and an unmaintainable project.

Since we have had similar feature requests:

- #3411
- #3860
- #4125

we had to design a new feature that met people's needs before implementing some of the requirements. Every issue above is related to how to determine lint target files.

**Comment by silverwind** [comment]
Reading through the [eslint flat config changes](https://eslint.org/blog/2022/08/new-config-system-part-2/), it does seem like a complete rewrite of the config system. They moved/removed some options, got rid of the `overrides` mechanism etc, so it's one big breaking change that is governed by the move to the new config filename `eslint.config.js`.

So if stylelint were to adopt some of these ideas, I'd recommend moving to `stylelint.config.js` as well, so users can opt into the new config with the new filename. I definitely feel like having a single `.js` config is the way to go. That way you can eventually drop a few dependencies as well like the YAML parser.

**Comment by jeddy3** [comment]
One of the benefits of moving behind ESLint is that we can learn from their corrections. For example, we chose to add the `overrides` property because it [looked forward to one of the _underlying_ mechanisms of a flat config](https://eslint.org/docs/latest/use/configure/configuration-files-new#cascading-configuration-objects).

As @ybiquitous mentioned in https://github.com/stylelint/stylelint/issues/6689#issuecomment-1454743629, we have a few historical issues in this area (some from the time when we had out-of-the-box support for syntaxes like SCSS). Let's choose the approach that is forward-looking to a flat config.

When we eventually move to single `stylelint.config.js`, it'll be:

- easier for our users to migrate.
- less of a rewrite for us

Am I right in thinking that a top-level `files` is a forward-looking option because it'd sit alongside `rules` and `ignoreFiles` as an _implicit_ [configuration object ](https://eslint.org/docs/latest/use/configure/configuration-files-new#configuration-objects)?

**Comment by ybiquitous** [comment]
> Am I right in thinking that a top-level `files` is a forward-looking option

Yeah, I think so.

**Comment by Mouvedia** [comment]
Not many of the subscribers on the original issue will see this discussion, due to the split. Let's give them time to chime in and voice their opinions.

**Comment by ybiquitous** [comment]
> Not many of the subscribers on the original issue will see this discussion, due to the split.

Unfortunately, I agree. 😓 

Considering the issue title, it seems we should discuss the `files` config property on #3860. Many interesting views have been presented already on this issue, though.

**Comment by jeddy3** [comment]
> it seems we should discuss the files config property on https://github.com/stylelint/stylelint/issues/3860.

I agree. We can close this issue.

We should bare in mind that a lot of our historical issues:

- have only accrued thumbs up because they have been open for a while
- may now be redundant because we're added another feature that solves the problem in another way

At some point, we should go through our historical issues and close the now-irrelevant ones. Unfortunately, I suspect we all lack time to do that.

**Comment by ybiquitous** [comment]
I agree with closing this issue to direct the discussion to #3860. If we need the `--ext` flag again after the discussion, then let's reopen this.

> At some point, we should go through our historical issues and close the now-irrelevant ones. Unfortunately, I suspect we all lack time to do that.

Yeah, totally agree. I will try triaging old issues as possible if I notice them.

**Comment by silverwind** [comment]
Assuming you go for more breaking changes later with a new config format, I would suggest to still merge https://github.com/stylelint/stylelint/pull/6617 now and then deprecate/remove it later with the config rewrite.

**Comment by jeddy3** [comment]
We'll want to be forward facing so that user migration is more straightforward. It's why we added `overrides`.

If we add `files` as a top-level property, then the current config would be:

```json
{
  "files": "*.scss",
  "rules": {
    "unit-allowed-list": ["rem"]
  },
  "overrides": [
    {
      "files": "**/components/*.scss"
      "rules": {
        "function-disallowed-list": ["calc"]
      }
    }
  ]
}
```

Which can be run with:

```console
npx stylelint
```

If we then move to a flat config, the migration will only be syntactic:

```js
export default [
  {
    "files": "*.scss"
    "rules": {
      "unit-allowed-list": ["rem"]
    }
  },
  {
    "files": "**/components/*.scss"
    "rules": {
      "function-disallowed-list": ["calc"]
    }
  }
]
```

If we add `--ext`, it's a different mechanism they'll need to migrate from.

The suggestion is to close this issue and https://github.com/stylelint/stylelint/pull/6617, and pick up https://github.com/stylelint/stylelint/issues/3860 instead.

**Comment by silverwind** [comment]
@jeddy3 sounds good. I would only suggest making `files` accept an array of strings instead of just single strings. This is also matching the eslint implementation and is much more flexible.

**Comment by ybiquitous** [comment]
@jeddy3 I totally agree with your suggestion <https://github.com/stylelint/stylelint/issues/6689#issuecomment-1465130308>.

@silverwind Yes, `files` will support a string or an array of strings like `overrides[].files`.

https://github.com/stylelint/stylelint/blob/5d46c7dcc8e4516f9c8ef2f8509c1702d4cfed1a/types/stylelint/index.d.ts#L12-L14

**Comment by silverwind** [comment]
> Yes, files will support a string or an array of strings like overrides[].files.

It's a very minor detail but ESLint Flat only supports the array form for both `files` and `overrides`. I think it makes sense because plural implies multiple items to me.

**Comment by ybiquitous** [comment]
> I think it makes sense because plural implies multiple items to me.

I have no strong opinions, but other properties support `string | string[]`:

https://github.com/stylelint/stylelint/blob/5d46c7dcc8e4516f9c8ef2f8509c1702d4cfed1a/types/stylelint/index.d.ts#L5-L7

Considering consistency and backward compatibility, it seems to make sense that the `files` property also supports `string | string[]`. Anyway, let's pick it up again on #3860.

**Comment by Mouvedia** [comment]
Since no one came in support for this ticket, the PR has been closed.

**Comment by ybiquitous** [comment]
> I cannot remove the _status: wip_ label.

Done.

**Comment by jeddy3** [comment]
Closing in favour of https://github.com/stylelint/stylelint/issues/3860

**Comment by Mouvedia** [comment]
@jeddy3 should we reevaluate in light of https://eslint.org/blog/2025/02/eslint-v9.21.0-released/#--ext-cli-option?
